pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/BaseMailboxReceiveOperator.java

```diff
@@ -90,9 +106,108 @@ public void cancel(Throwable t) {
   }
 
   protected void cancelRemainingMailboxes() {
-    ReceivingMailbox mailbox;
-    while ((mailbox = _mailboxes.poll()) != null) {
+    for (ReceivingMailbox mailbox : _mailboxes) {
       mailbox.cancel();
     }
   }
+
+  public void onData() {
+    _lock.lock();
+    try {
+      _notEmpty.signal();
+    } finally {
+      _lock.unlock();
+    }
+  }
+
+  protected TransferableBlock readBlockBlocking() {
```

Comment:
I've move this code here because both ReceivingMailboxOperator and SortedReceivingMailboxOperator use the same code.